examples/curves_analysis.rs

```diff
@@ -0,0 +1,149 @@
+        order_minus_one + 1
+    };
+
+    {
+        let prob = rug::Rational::from((
+            rug::Integer::from(rug::Integer::ONE << 256) - &order,
+            rug::Integer::from(rug::Integer::ONE << 256),
+        ));
+        println!(
+            "Probability that random 32 bytes are invalid scalar: {prob_num} in 2^256: {prob_f64:.4} in range {prob_range}",
+            prob_num = prob.numer(),
+            prob_f64 = prob.to_f64(),
+            prob_range = find_closest_power_of_minus_two_str(&prob),
+        );
+    }
+
+    println!("order = {order}");
+
+    for rand_bytes in rand_bytes {
+        let rand_bits = rand_bytes * 8;
+        let rand_mod: rug::Integer = (rug::Integer::ONE << rand_bits).into();
+
+        let (n, w) = rand_mod.div_rem_ref(&order).complete();
+
+        let prob_uniform = rug::Rational::from((rug::Integer::ONE, &order));
+        println!("When we take rand({rand_bytes} bytes) modulo curve_order, we have distribution:");
+        {
+            let prob_actual = rug::Rational::from((&n + 1, &rand_mod));
+            println!(
+                "- for any y < w, Pr[x = y] = {} in 2^{rand_bits}, diff with uniform in range: {range}",
+                rug::Integer::from(&n + 1),
+                range = find_closest_power_of_minus_two_str(&(prob_actual - &prob_uniform).abs())
+            );
+        }
+        {
+            let prob_actual = rug::Rational::from((&n, &rand_mod));
+            println!(
+                "- for any y >= w, Pr[x = y] = {n} in 2^{rand_bits}, diff with uniform in range: {range}",
+                range = find_closest_power_of_minus_two_str(&(prob_actual - &prob_uniform).abs())
+            );
+        }
+        println!("where w = {w}");
+
+        let stat_diff = {
+            let before_w_per_each =
+                (rug::Rational::from((&n + 1, &rand_mod)) - &prob_uniform).abs();
+            let before_w = &w * before_w_per_each;
+
+            let after_w_per_each = (rug::Rational::from((&n, &rand_mod)) - &prob_uniform).abs();
+            let after_w = (&order - &w).complete() * after_w_per_each;
+
+            (before_w + after_w) / 2
+        };
+
+        println!(
+            "statistical difference between actual and uniform distribution is in range: {}",
+            find_closest_power_of_minus_two_str(&stat_diff)
+        );
+
+        assert_eq!(stat_diff, stat_diff_from_book(&rand_mod, &order));
+    }
+}
+
+/// Given 0 < x < 1, returns `lo`, `hi` such that 2^-lo <= x <= 2^-hi
+fn find_closest_power_of_minus_two(x: &rug::Rational) -> (u32, u32) {
+    assert!(x.is_positive() && x < rug::Rational::ONE);
+
+    for (lo, hi) in (1..).zip(0..) {
+        let x_lo = rug::Rational::from((rug::Integer::ONE, rug::Integer::ONE << lo));
+        let x_hi = rug::Rational::from((rug::Integer::ONE, rug::Integer::ONE << hi));
+        if x_lo <= *x && *x <= x_hi {
+            return (lo, hi);
+        }
+    }
```

Comment:
You don't need two variables here: as soon as x_lo <= x you can take hi = lo - 1

Reply:
What you suggest is supposed to decrease amount of the code, but adding `let hi = lo - 1` actually adds extra line of code

Reply:
(if I correctly understood what you suggested)

Reply:
No, I'm suggesting reducing the amount of computation done, not necessarily code. You remove hi in the loop, you remove x_hi altogether, and when the comparison succeeds, you compute `hi = lo - 1`. By my calculations, the amount of lines of code will stay the same, but the lines will get shorter and we drop one comparison

Reply:
if $2^{-k} \le x$, that does not necessarily mean that $x \le 2^{-k+1}$ in general. That’s probably true from the fact that we check that $0 < x < 1$, but again, in this tool I prefer to double check everything, because we do complex computations and it’s essential to be assured that everything is correct. I prioritize auditability/readability here over performance/allocations

Reply:
In general no, but the loop finds not a general `k`, but a minimal k such that $x \le 2^{-k}$. Since it's minimal, it follows that $x > 2^{-(k - 1)}$.

Reformulating imperatively, you're iterating `lo <- [1, 2, ..]`, you compute `x_lo = 1 / (1 << lo)`. You condition for breaking is `x_lo <= x`. that means if you're evaluating the loop, you have `x_lo' = 1 / (1 << (lo - 1))` and `x_lo' > x`. Note that this x_lo' is exactly x_hi.

What it means is that you can skip comparing x_hi, and move computing x_hi altogether into the same branch as return.

Reply:
I get your view about not optimizing much for performance, since it's a piece of code ran rarely. But also when I first saw it, it took me some time to verify that `hi` is indeed computed from k-1, so readability gain here is dubious

Reply:
Here's my logic: when I see in the output of the program that "stat diff lies in $2^{-k} \le x \le 2^{-k+1}$", my first thought is to check if it's correct. Suppose $x$ is calculated correctly, I need to check if boundaries are correct as well. To do that, I have to analyze the function. Right now, it's very easy to check that boundaries are identified correctly, as the only way this function returns something is when the condition $2^{-k} \le x \le 2^{-k+1}$ is true:

https://github.com/LFDT-Lockness/hd-wallet/blob/158b65ce54ed6d0d031a9414890b4a9170dd4f11/examples/curves_analysis.rs#L126-L130

So that's very easy to audit and analyze. If we do what you suggested, in my opinion, it will be more difficult to analyze the function, as it's now required to understand how the loop works. As you said, in general, this approach does not yield correct result, but it does in this particular case, and it will take some time to come to the same conclusion.

In this case, I want to keep the option that looks correct right away,

Reply:
> But also when I first saw it, it took me some time to verify that hi is indeed computed from k-1

How come? It's very clearly defined in the loop that `lo` goes in `1..` while `hi` goes in `0..`